src/lua.rs

```diff
@@ -625,6 +637,59 @@ impl Lua {
         }
     }
 
+    /// Set a function for Lua to call on conditions configured by `triggers`. This function will
+    /// always succeed, but the hook may raise an error to be reported back to the caller when
+    /// running Lua code.
+    ///
+    /// You may use this functionality for a few reasons. You may limit yourself to viewing the
+    /// information or react to certain events. However, you can also use it to limit for how long
+    /// the script will run.
+    ///
+    /// # Example
+    ///
+    /// Shows each line number of code being executed by the Lua interpreter.
+    ///
+    /// ```
+    /// # extern crate rlua;
+    /// # use rlua::{Lua, HookTriggers, Debug};
+    /// # fn main() {
+    /// let code = r#"local x = 2 + 3
+    /// local y = x * 63
+    /// local z = string.len(x..", "..y)
+    /// "#;
+    ///
+    /// let lua = Lua::new();
+    /// lua.set_hook(HookTriggers {
+    ///     every_line: true, ..Default::default()
+    /// }, |_lua: Lua, debug: &Debug| {
+    ///     println!("line {}", debug.curr_line);
+    ///     Ok(())
+    /// });
+    /// let _ = lua.exec::<_, ()>(code, None);
+    /// # }
+    /// ```
+    pub fn set_hook<F>(
+        &self,
+        triggers: HookTriggers,
+        callback: F)
+    where
+        F: 'static + Send + FnMut(Lua, &Debug) -> Result<()>
+    {
+        unsafe {
+            (*extra_data(self.state)).hook_callback = Some(Rc::new(RefCell::new(callback)));
+            ffi::lua_sethook(self.state, Some(hook_proc), triggers.mask(), triggers.count());
+        }
+    }
+
+    /// Remove any hook previously set by `set_hook`. This function has no effect if no hook was
+    /// set.
+    pub fn remove_hook(&self) {
+        unsafe {
+            (*extra_data(self.state)).hook_callback = None;
+            ffi::lua_sethook(self.state, None, 0, 0);
+        }
+    }
```

Comment:
The hook callback function should take a `&Lua` otherwise this is unsound.

Reply:
Can't remember if I did the change. Might have forgotten to push. I passed it by value since I thought it would not be used anyway in `hook_proc`.